xmtp_mls/README.md

```diff
@@ -0,0 +1,136 @@
+    sent_at_ns BIGINT NOT NULL, -- Based on the timestamp of the message
+    sender_installation_id TEXT NOT NULL,
+    sender_wallet_address TEXT NOT NULL, -- Could remove this if we added a table mapping installation_ids to wallet addresses
+}
+
+-- Used to keep track of the last seen message timestamp in a topic
+CREATE TABLE topic_refresh_state {
+    topic TEXT PRIMARY KEY NOT NULL,
+    last_message_timestamp BIGINT NOT NULL
+}
+
+-- This table is required to retry messages that do not send successfully due to epoch conflicts
+CREATE TABLE group_intents {
+    id SERIAL PRIMARY KEY NOT NULL,
+    kind INT NOT NULL, -- INTENT_KIND
+    group_id TEXT NOT NULL,
+    "data" BLOB NOT NULL, -- Some sort of serializable blob that can be used to re-try the message if the first attempt failed due to conflict
+    "state" INT NOT NULL, -- INTENT_STATE,
+    message_hash TEXT -- The hash of the encrypted, concrete, form of the message if it was published.
+}
+
+CREATE TABLE pending_outbound_welcome_messages {
+    id TEXT PRIMARY KEY NOT NULL, -- Derived via SHA256(CONCAT(group_id, welcome_message, ratchet_tree, installation_id))
+    installation_id TEXT PRIMARY KEY NOT NULL,
+    commit_hash TEXT NOT NULL, -- The hash of the commit which created this
+    group_id TEXT NOT NULL,
+    welcome_message BLOB NOT NULL,
+    ratchet_tree BLOB NOT NULL,
+}
+
+-- OpenMLS Keystore backing database to lookup objects. Interface and matching types defined here https://github.com/openmls/openmls/blob/main/traits/src/key_store.rs
+CREATE TABLE openmls_keystore {
+    "key" BLOB PRIMARY KEY NOT NULL,
+    "value" BLOB NOT NULL,
+}
+```
+
+## Enums
+
+### GROUP_MEMBERSHIP_STATE
+
+- ALLOWED // User has agreed to be a member of the group
+- REJECTED // User has rejected an invite to the group or left
+- PENDING // User has neither accepted or rejected whether they should join the group
+
+### INTENT_STATE
+
+- TO_SEND // Either has never been sent to the network or needs to be re-sent
+- PUBLISHED // Sent to the network but has not been read back or committed
+- COMMITTED // Committed messages could be deleted
+
+### INTENT_KIND
+
+- APPLICATION // An intent to send a message to the group
+- ADD_MEMBERS // An intent to add members to the group
+- REMOVE_MEMBERS // An intent to remove members from the group
+- KEY_UPDATE // An intent to update your own group key
+
+## State Machine
+
+The [following diagram](https://app.excalidraw.com/s/4nwb0c8ork7/6pPH1kQDoj3) illustrates some common flows in the state machine
+
+![MLS State Machione](../img/mls-state-machine.png "MLS State Machine")
+
+### Known missing items from the state machine
+
+- Key updates
+- Processing incoming welcome messages
+
+### Add members to a group
+
+Simplified high level flow for adding members to a group:
+
+1. Consume Key Packages for all new members
+1. Create a `group_intent` for adding the members
+1. Sync the state of the group with the network
```

Comment:
1. We don't strictly need to sync the state. It just makes conflicts less likely. Given that adds/removes are relatively rare it might make sense. Also might be totally overkill.
2. I mean process all the unread messages on the group topic
3. I've left subscriptions out of the equation here, but I have been thinking about them a bit. Subscriptions don't guarantee ordering or completeness of results, and that's unlikely to change. For us to support subscriptions in MLS I think the flow needs to be:
- Open a subscription to the group topic
- For any application message for an epoch you have keys for, process it immediately
- When you receive a commit, trigger a call to the network to read from the group topic in the committed order. Discard any application messages you have already saved, and process all commits in order.

That lets us have real-time messaging without negating the benefits of total ordering we get from the query API.